src/particles/diagnostics/ReducedBeamCharacteristics.cpp

```diff
@@ -403,7 +406,183 @@ namespace impactx::diagnostics
+        amrex::ParticleReal const beta_y = y_msd / emittance_yd;
+        amrex::ParticleReal const beta_t = t_ms / emittance_t;
+        // Courant-Snyder (Twiss) alpha
+        amrex::ParticleReal const alpha_x = - xpx_d / emittance_xd;
+        amrex::ParticleReal const alpha_y = - ypy_d / emittance_yd;
+        amrex::ParticleReal const alpha_t = - tpt / emittance_t;
+
+        // Calculate normalized emittances
+        amrex::ParticleReal emittance_xn = emittance_x * bg;
+        amrex::ParticleReal emittance_yn = emittance_y * bg;
+        amrex::ParticleReal emittance_tn = emittance_t * bg;
+
+        // Determine whether to calculate eigenemittances, and initialize
+        amrex::ParmParse pp_diag("diag");
+        bool compute_eigenemittances = false;
+        pp_diag.queryAdd("eigenemittances", compute_eigenemittances);
+        amrex::ParticleReal emittance_1 = emittance_xn;
+        amrex::ParticleReal emittance_2 = emittance_yn;
+        amrex::ParticleReal emittance_3 = emittance_tn;
+
+        if (compute_eigenemittances) {
+           // Store the covariance matrix in dynamical variables:
+           amrex::SmallMatrix<amrex::ParticleReal, 6, 6, amrex::Order::F, 1> Sigma;
+           Sigma(1,1) = x_ms;
+           Sigma(1,2) = xpx * bg;
+           Sigma(1,3) = xy;
+           Sigma(1,4) = xpy * bg;
+           Sigma(1,5) = xt;
+           Sigma(1,6) = xpt * bg;
+           Sigma(2,1) = xpx * bg;
+           Sigma(2,2) = px_ms * bg2;
+           Sigma(2,3) = pxy * bg;
+           Sigma(2,4) = pxpy * bg2;
+           Sigma(2,5) = pxt * bg;
+           Sigma(2,6) = pxpt * bg2;
+           Sigma(3,1) = xy;
+           Sigma(3,2) = pxy * bg;
+           Sigma(3,3) = y_ms;
+           Sigma(3,4) = ypy * bg;
+           Sigma(3,5) = yt;
+           Sigma(3,6) = ypt * bg;
+           Sigma(4,1) = xpy * bg;
+           Sigma(4,2) = pxpy * bg2;
+           Sigma(4,3) = ypy * bg;
+           Sigma(4,4) = py_ms * bg2;
+           Sigma(4,5) = pyt * bg;
+           Sigma(4,6) = pypt * bg2;
+           Sigma(5,1) = xt;
+           Sigma(5,2) = pxt * bg;
+           Sigma(5,3) = yt;
+           Sigma(5,4) = pyt * bg;
+           Sigma(5,5) = t_ms;
+           Sigma(5,6) = tpt * bg;
+           Sigma(6,1) = xpt * bg;
+           Sigma(6,2) = pxpt * bg2;
+           Sigma(6,3) = ypt * bg;
+           Sigma(6,4) = pypt * bg2;
+           Sigma(6,5) = tpt * bg;
+           Sigma(6,6) = pt_ms * bg2;
+           // Calculate eigenemittances
+           std::tuple<amrex::ParticleReal, amrex::ParticleReal, amrex::ParticleReal> emittances = Eigenemittances(Sigma);
+           emittance_1 = std::get<0>(emittances);
+           emittance_2 = std::get<1>(emittances);
+           emittance_3 = std::get<2>(emittances);
+        }
+
+        auto const nan = std::numeric_limits<amrex::ParticleReal>::quiet_NaN();
+
+        std::unordered_map<std::string, amrex::ParticleReal> data;
+        data["x_mean"] = nan;
```

Comment:
We can check if most of those are truly undefined or instead are just "trivial" values (like `0`).